d2ir/d2ir.go

```diff
@@ -825,65 +896,235 @@ func (m *Map) GetEdges(eid *EdgeID) []*Edge {
+		fa, err := m.EnsureField(commonKP, nil, true)
+		if err != nil {
+			return err
 		}
-		if f.Map() == nil {
-			f.Composite = &Map{
-				parent: f,
+		for _, f := range fa {
+			if _, ok := f.Composite.(*Array); ok {
+				return d2parser.Errorf(refctx.Edge.Src, "cannot index into array")
+			}
+			if f.Map() == nil {
+				f.Composite = &Map{
+					parent: f,
+				}
+			}
+			err = f.Map().createEdge(eid, refctx, ea)
+			if err != nil {
+				return err
 			}
 		}
-		return f.Map().CreateEdge(eid, refctx)
+		return nil
 	}
 
 	ij := findProhibitedEdgeKeyword(eid.SrcPath...)
 	if ij != -1 {
-		return nil, d2parser.Errorf(refctx.Edge.Src.Path[ij].Unbox(), "reserved keywords are prohibited in edges")
+		return d2parser.Errorf(refctx.Edge.Src.Path[ij].Unbox(), "reserved keywords are prohibited in edges")
 	}
 	ij = findBoardKeyword(eid.SrcPath...)
 	if ij == len(eid.SrcPath)-1 {
-		return nil, d2parser.Errorf(refctx.Edge.Src.Path[ij].Unbox(), "edge with board keyword alone doesn't make sense")
-	}
-	src := m.GetField(eid.SrcPath...)
-	if NodeBoardKind(src) != "" {
-		return nil, d2parser.Errorf(refctx.Edge.Src, "cannot create edges between boards")
+		return d2parser.Errorf(refctx.Edge.Src.Path[ij].Unbox(), "edge with board keyword alone doesn't make sense")
 	}
 
 	ij = findProhibitedEdgeKeyword(eid.DstPath...)
 	if ij != -1 {
-		return nil, d2parser.Errorf(refctx.Edge.Dst.Path[ij].Unbox(), "reserved keywords are prohibited in edges")
+		return d2parser.Errorf(refctx.Edge.Dst.Path[ij].Unbox(), "reserved keywords are prohibited in edges")
 	}
 	ij = findBoardKeyword(eid.DstPath...)
 	if ij == len(eid.DstPath)-1 {
-		return nil, d2parser.Errorf(refctx.Edge.Dst.Path[ij].Unbox(), "edge with board keyword alone doesn't make sense")
+		return d2parser.Errorf(refctx.Edge.Dst.Path[ij].Unbox(), "edge with board keyword alone doesn't make sense")
+	}
+
+	srcKP := d2ast.MakeKeyPath(eid.SrcPath)
+	lastMatch := 0
+	for i, el := range srcKP.Path {
+		for j := lastMatch; j < len(refctx.Edge.Src.Path); j++ {
+			realEl := refctx.Edge.Src.Path[j]
+			if el.ScalarString() == realEl.ScalarString() {
+				srcKP.Path[i] = realEl
+				lastMatch += j + 1
+			}
+		}
+	}
+	dstKP := d2ast.MakeKeyPath(eid.DstPath)
+	lastMatch = 0
+	for i, el := range dstKP.Path {
+		for j := lastMatch; j < len(refctx.Edge.Dst.Path); j++ {
+			realEl := refctx.Edge.Dst.Path[j]
+			if el.ScalarString() == realEl.ScalarString() {
+				dstKP.Path[i] = realEl
+				lastMatch += j + 1
+			}
+		}
+	}
+
+	srcFA, err := m.EnsureField(srcKP, nil, true)
+	if err != nil {
+		return err
+	}
+	dstFA, err := m.EnsureField(dstKP, nil, true)
+	if err != nil {
+		return err
+	}
+
+	for _, src := range srcFA {
+		for _, dst := range dstFA {
+			if src == dst && (len(srcFA) > 1 || len(dstFA) > 1) {
+				// Globs do not make self edges.
+				continue
+			}
+
+			if srcKP.HasDoubleGlob() || dstKP.HasDoubleGlob() {
+				// If either has a double glob we only select leafs, those without children.
```

Comment:
it shouldn't be either, it should only apply to the side that has the double glob.

Otherwise you can't do this:

```d2
fast: {
  a
  far
}

task: {
  a
}

task.** -> fast
```

![Screen Shot 2023-07-30 at 11 33 02 AM](https://github.com/terrastruct/d2/assets/3120367/abb36050-c06f-4daf-905e-25dff5e8f4fe)

Reply:
fixed.